File: lib/billing/provider-settings.ts
import { db } from "@/lib/db";
import { systemSettings } from "@/lib/db/schema";
import { eq } from "drizzle-orm";

export const PROVIDER_KEYS = {
  STRIPE_SECRET_KEY: "stripe_secret_key",
  STRIPE_WEBHOOK_SECRET: "stripe_webhook_secret",
  XENDIT_SECRET_KEY: "xendit_secret_key",
  XENDIT_WEBHOOK_TOKEN: "xendit_webhook_token",
  LEMONSQUEEZY_API_KEY: "lemonsqueezy_api_key",
  LEMONSQUEEZY_STORE_ID: "lemonsqueezy_store_id",
  LEMONSQUEEZY_WEBHOOK_SECRET: "lemonsqueezy_webhook_secret",
  EXTERNAL_TAX_PROVIDER: "external_tax_provider",
  TAXJAR_API_KEY: "taxjar_api_key",
} as const;

type ProviderKey = (typeof PROVIDER_KEYS)[keyof typeof PROVIDER_KEYS];

const cache = new Map<string, { value: string; expiresAt: number }>();
const CACHE_TTL = 60_000;

export async function getProviderSetting(key: ProviderKey): Promise<string> {
  const cached = cache.get(key);
  if (cached && cached.expiresAt > Date.now()) {
    return cached.value;
  }

  try {
    const [row] = await db
      .select()
      .from(systemSettings)
      .where(eq(systemSettings.key, key));

    if (row && row.value) {
      cache.set(key, { value: row.value, expiresAt: Date.now() + CACHE_TTL });
      return row.value;
    }
  } catch {
    // no-op
  }

  const envKey = key.toUpperCase();
  return process.env[envKey] ?? "";
}

export async function getProviderStatus(): Promise<{
  stripe: { configured: boolean; secretKey: string; webhookSecret: string };
  xendit: { configured: boolean; secretKey: string; webhookToken: string };
  lemonsqueezy: { configured: boolean; apiKey: string; storeId: string; webhookSecret: string };
  tax: { configured: boolean; externalProvider: string; taxjarApiKey: string };
}> {
  const [
    stripeKey, stripeWebhook,
    xenditKey, xenditToken,
    lsKey, lsStore, lsWebhook,
    externalTaxProvider, taxjarApiKey,
  ] = await Promise.all([
    getProviderSetting(PROVIDER_KEYS.STRIPE_SECRET_KEY),
    getProviderSetting(PROVIDER_KEYS.STRIPE_WEBHOOK_SECRET),
    getProviderSetting(PROVIDER_KEYS.XENDIT_SECRET_KEY),
    getProviderSetting(PROVIDER_KEYS.XENDIT_WEBHOOK_TOKEN),
    getProviderSetting(PROVIDER_KEYS.LEMONSQUEEZY_API_KEY),
    getProviderSetting(PROVIDER_KEYS.LEMONSQUEEZY_STORE_ID),
    getProviderSetting(PROVIDER_KEYS.LEMONSQUEEZY_WEBHOOK_SECRET),
    getProviderSetting(PROVIDER_KEYS.EXTERNAL_TAX_PROVIDER),
    getProviderSetting(PROVIDER_KEYS.TAXJAR_API_KEY),
  ]);

  return {
    stripe: {
      configured: !!stripeKey,
      secretKey: maskValue(stripeKey),
      webhookSecret: maskValue(stripeWebhook),
    },
    xendit: {
      configured: !!xenditKey,
      secretKey: maskValue(xenditKey),
      webhookToken: maskValue(xenditToken),
    },
    lemonsqueezy: {
      configured: !!lsKey && !!lsStore,
      apiKey: maskValue(lsKey),
      storeId: lsStore || "",
      webhookSecret: maskValue(lsWebhook),
    },
    tax: {
      configured: !!externalTaxProvider,
      externalProvider: externalTaxProvider || "",
      taxjarApiKey: maskValue(taxjarApiKey),
    },
  };
}

function maskValue(value: string): string {
  if (!value) return "";
  if (value.length <= 8) return "••••••••";
  return "••••••••" + value.slice(-4);
}

export async function saveProviderSetting(
  key: ProviderKey,
  value: string,
  sensitive = true,
): Promise<void> {
  await db
    .insert(systemSettings)
    .values({ key, value, sensitive, updatedAt: new Date() })
    .onConflictDoUpdate({
      target: systemSettings.key,
      set: { value, sensitive, updatedAt: new Date() },
    });

  cache.delete(key);
}

export function clearProviderCache(): void {
  cache.clear();
}

File: rustbill/crates/rustbill-core/src/settings/provider_settings.rs
//! Payment provider settings: DB storage with in-memory cache + env fallback.

use sqlx::PgPool;
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::RwLock;

const CACHE_TTL: Duration = Duration::from_secs(60);

pub struct ProviderSettingsCache {
    cache: RwLock<HashMap<String, CacheEntry>>,
    pool: PgPool,
}

struct CacheEntry {
    value: String,
    expires_at: Instant,
}

/// A snapshot of provider settings for use in payment operations.
#[derive(Debug, Clone, Default)]
pub struct ProviderSettings {
    settings: HashMap<String, String>,
}

impl ProviderSettings {
    pub fn get(&self, key: &str) -> Option<String> {
        self.settings.get(key).cloned()
    }
}

impl ProviderSettingsCache {
    pub fn new(pool: PgPool) -> Arc<Self> {
        Arc::new(Self {
            cache: RwLock::new(HashMap::new()),
            pool,
        })
    }

    /// Get a single setting value: cache -> DB -> env fallback.
    pub async fn get(&self, key: &str) -> String {
        // Check cache
        {
            let cache = self.cache.read().await;
            if let Some(entry) = cache.get(key) {
                if entry.expires_at > Instant::now() {
                    return entry.value.clone();
                }
            }
        }

        // Try DB
        let db_value: Option<String> =
            sqlx::query_scalar("SELECT value FROM system_settings WHERE key = $1")
                .bind(key)
                .fetch_optional(&self.pool)
                .await
                .ok()
                .flatten();

        if let Some(ref val) = db_value {
            let mut cache = self.cache.write().await;
            cache.insert(
                key.to_string(),
                CacheEntry {
                    value: val.clone(),
                    expires_at: Instant::now() + CACHE_TTL,
                },
            );
            return val.clone();
        }

        // Env fallback
        let env_key = key.to_uppercase();
        let env_val = std::env::var(&env_key).unwrap_or_default();
        env_val
    }

    /// Get all settings for a provider as a ProviderSettings snapshot.
    pub async fn get_provider_settings(&self, keys: &[&str]) -> ProviderSettings {
        let mut settings = HashMap::new();
        for key in keys {
            let val = self.get(key).await;
            if !val.is_empty() {
                settings.insert(key.to_string(), val);
            }
        }
        ProviderSettings { settings }
    }

    /// Save a setting to DB and invalidate cache.
    pub async fn save(&self, key: &str, value: &str, sensitive: bool) -> crate::error::Result<()> {
        sqlx::query(
            r#"INSERT INTO system_settings (key, value, sensitive, updated_at)
            VALUES ($1, $2, $3, NOW())
            ON CONFLICT (key) DO UPDATE SET value = $2, sensitive = $3, updated_at = NOW()"#,
        )
        .bind(key)
        .bind(value)
        .bind(sensitive)
        .execute(&self.pool)
        .await?;

        let mut cache = self.cache.write().await;
        cache.remove(key);
        Ok(())
    }

    /// Get provider status with masked values for display.
    pub async fn get_status(&self) -> ProviderStatus {
        let stripe_key = self.get("stripe_secret_key").await;
        let stripe_webhook = self.get("stripe_webhook_secret").await;
        let xendit_key = self.get("xendit_secret_key").await;
        let xendit_token = self.get("xendit_webhook_token").await;
        let ls_key = self.get("lemonsqueezy_api_key").await;
        let ls_store = self.get("lemonsqueezy_store_id").await;
        let ls_webhook = self.get("lemonsqueezy_webhook_secret").await;

        ProviderStatus {
            stripe: ProviderInfo {
                configured: !stripe_key.is_empty(),
                secret_key: mask_value(&stripe_key),
                webhook_secret: mask_value(&stripe_webhook),
            },
            xendit: ProviderInfo {
                configured: !xendit_key.is_empty(),
                secret_key: mask_value(&xendit_key),
                webhook_secret: mask_value(&xendit_token),
            },
            lemonsqueezy: LsProviderInfo {
                configured: !ls_key.is_empty() && !ls_store.is_empty(),
                api_key: mask_value(&ls_key),
                store_id: ls_store,
                webhook_secret: mask_value(&ls_webhook),
            },
        }
    }

    pub async fn clear_cache(&self) {
        let mut cache = self.cache.write().await;
        cache.clear();
    }
}

fn mask_value(value: &str) -> String {
    if value.is_empty() {
        return String::new();
    }
    if value.len() <= 8 {
        return "••••••••".to_string();
    }
    format!("••••••••{}", &value[value.len() - 4..])
}

#[derive(Debug, serde::Serialize)]
pub struct ProviderStatus {
    pub stripe: ProviderInfo,
    pub xendit: ProviderInfo,
    pub lemonsqueezy: LsProviderInfo,
}

#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderInfo {
    pub configured: bool,
    pub secret_key: String,
    pub webhook_secret: String,
}

#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LsProviderInfo {
    pub configured: bool,
    pub api_key: String,
    pub store_id: String,
    pub webhook_secret: String,
}
